### PR TITLE
temporal: t.rast.aggregate: add -e flag to extend existing STRDS

### PR DESCRIPTION
**Pull request by Sourish-spc**
## Description

This PR adds support for extending an existing Space Time Raster Dataset (STRDS) in `t.rast.aggregate` by introducing a new `-e` flag, following the same pattern implemented in #3798 for `t.rast.neighbors`.

## Changes

- Added `-e` flag (`Extend existing space time raster dataset`) to module interface
- Modified STRDS creation logic to open existing STRDS when `-e` flag is set instead of always creating a new one
- Added history update via `update_command_string()` when extending existing STRDS
- Added test `test_extend_existing_strds` to existing test suite

## Usage

First run — creates new STRDS:
  t.rast.aggregate input=daily_rain output=monthly_rain granularity="1 month" method=average basename=result

Second run — extends existing STRDS:
  t.rast.aggregate input=daily_rain output=monthly_rain granularity="1 month" method=average basename=result_new -e

## Related

Closes part of #3427
See also #3798

### CODE REVIEW

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 1]
```diff
@@ -144,8 +151,24 @@ def main():
...
+    output_exists = output_strds.is_in_db(dbif)
+
+    if extend and not output_exists:
+        gs.fatal(
+            _(
+                "Space time raster dataset <%s> does not exist. "
+                "Cannot extend a non-existing dataset."
+            )
+            % output
+        )
```
Comment (ninsbl) [inline temporal/t.rast.aggregate/t.rast.aggregate.py]: Personally, I do not think a fatal error should be thrown here. I would say, a warning at best. In any case, this part should be consistent across all temporal tools that receive the e-flag.

**Review by ninsbl — CHANGES_REQUESTED** [review]
Please also install pre-commit in your local working directory to avoid CI runs for formalities...

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 1 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Updated: changed fatal error to warning and fall back to creating new STRDS when dataset doesn't exist.

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py [hunk 2]
```diff
@@ -237,6 +237,45 @@ def test_aggregation_3months(self):
...
+            granularity="1 month",
+            method="average",
+            sampling=["contains"],
+            overwrite=True,
+        )
+
+        # Verify STRDS was extended
+        lister = SimpleModule("t.rast.list", input="B", columns="name", flags="u")
+        self.runModule(lister)
+        self.assertIn("b_ext", lister.outputs.stdout)
```
Comment (wenzeslaus) [inline temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]: This only checks that some map with "b_ext" in the name exists. It doesn't verify the original maps are still present, that the STRDS wasn't just overwritten, and/or the total map count.

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 3]
```diff
@@ -226,9 +254,14 @@ def main():
         output_strds.set_aggregation_type(method)
         output_strds.metadata.update(dbif)
 
+        # Update command history if extending existing STRDS
+        if output_exists:
+            output_strds.update_command_string(dbif=dbif)
```
Comment (wenzeslaus) [inline temporal/t.rast.aggregate/t.rast.aggregate.py]: Does this make sense when overwriting?

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 4]
```diff
@@ -226,9 +254,14 @@ def main():
...
+
     dbif.close()
 
 
 if __name__ == "__main__":
     options, flags = gs.parser()
+
```
Comment (wenzeslaus) [inline temporal/t.rast.aggregate/t.rast.aggregate.py]: Please, review your code, including the PR changes.

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 5]
```diff
@@ -195,22 +219,26 @@ def main():
...
+                "strds",
+                temporal_type,
+                title,
+                description,
+                semantic_type,
+                dbif,
+                overwrite,
+            )
+        elif output_exists and extend:
+            output_strds = tgis.open_old_stds(output, "strds", dbif)
```
Comment (wenzeslaus) [inline temporal/t.rast.aggregate/t.rast.aggregate.py]: What happens when "else"? Or is there a condition which I don't see. Perhaps worth commenting on if that's the case.

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 6]
```diff
@@ -144,8 +151,25 @@ def main():
...
+
+    if extend and not output_exists:
+        gs.warning(
+            _(
+                "Space time raster dataset <%s> does not exist. "
+                "Cannot extend a non-existing dataset."
+            )
+            % output
+        )
+        extend = False  # fall back to creating new STRDS
```
Comment (wenzeslaus) [inline temporal/t.rast.aggregate/t.rast.aggregate.py]: The warning does not make sense to me. Either -e is "extend if already exist" (no error or warning needed) or "extend existing" (so error on non-existence). The first option is sort of append mode; we have that elsewhere.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 6 above)
Comment (ninsbl) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Agreed. That said, the confusion is my fault as I have not been explicit.
Personally, I would prefer "extend-if-exists" behavior, as it is more graceful, but I have no strong opinion here...

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 5 above)
Comment (ninsbl) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: I noticed that in _the.rast.neighbors_ too, where this is likely copied from. So It should be addressed there in a different PR too. Given that the possibility to extend existing STDS would be useful in multiple temporal tools, maybe a `extend_if_exists` (or something like that) library function would make sense, that returns either a new or exiting STDS!? @wenzeslaus what do you think?

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 3 above)
Comment (ninsbl) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Good point, iI did not really think about it. It may make sense in any case. I did not check the behavior. Updates of the STDS should be written to the history I`ld say, which is what _update_command_string_ does. The command should not be duplicated in the history though. That should be checked...

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 5 above)
Comment (ninsbl) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Maybe enhance and use https://github.com/OSGeo/grass/blob/65a1515640236673151f9c86e976e458a887197f/python/grass/temporal/open_stds.py#L91 ?

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 4 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Thank you for the review. The extra blank line has been removed in the latest commit.

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py (hunk 2 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]: Thank you for the review. The test has been strengthened to also verify that original maps are still present and the total map count increased after extending.

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 5 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Refactored the extend logic into check_new_stds in open_stds.py as a reusable library function with an extend parameter. The else case is now handled there: if -e is set and STRDS does not exist, it silently creates a new one (extend-if-exists behavior).

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 3 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Updated the condition to if not output_exists or extend to avoid duplicating the command string in history when overwriting, while still updating it correctly when extending.

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 6 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Removed the warning and fallback. Now using extend-if-exists behavior: if the STRDS doesn't exist, it just creates a new one silently.

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 5 above)
Comment (ninsbl) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Hi, an sorry for possibly putting you on a slightly wrong track with enhancing _check_new_stds_ . That function should probably stay as is. 
What in general is needed is 1) a function that checks if output STDS can be opened as requested (new or existing for update, e.g. _check_open_output_stds_). That should be minimal (not creating or opening) and run before main processing starts. And 2), after main processing succeeded, the output STDS should be opened or created so produced maps can be registered (e.g. _open_output_stds_).

Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py]: Thanks for the direction. I have reverted check_new_stds to its original state and created two new functions check_open_output_stds and open_output_stds in open_stds.py as you suggested. Also added them to __init__.py for export.

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 7]
```diff
@@ -144,8 +151,16 @@ def main():
...
+    # Check if output can be opened as requested before processing starts.
+    # This is a lightweight check that does not create or open the dataset.
+    tgis.check_open_output_stds(
+        output, "strds", dbif=dbif, overwrite=overwrite, extend=extend
+    )
+
+    # Record whether output already exists for use after processing
+    mapset = tgis.get_current_mapset()
+    output_id = output + "@" + mapset
+    output_exists = tgis.SpaceTimeRasterDataset(output_id).is_in_db(dbif)
```
Comment (ninsbl) [inline temporal/t.rast.aggregate/t.rast.aggregate.py]: This partially repeats the check you added above. If the results of the check are needed later on, the check_open_output_stds should return accordingly. It may be a pre-filled open fuction with functools.partial, for example, that is executed at the end or jus a boolean for output_exists...

File: python/grass/temporal/open_stds.py [hunk 8]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+    name,
+    type,
+    temporaltype,
+    title,
+    descr,
+    semantic,
+    dbif=None,
+    overwrite: bool = False,
+    extend: bool = False,
+):
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py:306]: Please add annotations.

File: python/grass/temporal/open_stds.py [hunk 9]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+    elif type in {"stvds", "vect", "vector"}:
+        sp = dataset_factory("stvds", id)
+    else:
+        msgr.fatal(_("Unknown type: %s") % (type))
+    if extend and sp.is_in_db(dbif):
+        # extend-if-exists: load and return existing dataset
+        sp.select(dbif)
+        if connection_state_changed:
+            dbif.close()
+        return sp
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py:357]: You duplicate quite a bit of code... You may consider adding a helper function similar to e.g. (untested, so not suitable for copy and paste!):

```
def get_stds(stds_id: str, stds_type: str) -> SpaceTimeDataset:
    """Return an initialized SpaceTimeDataset (STDS)."""
    msgr = get_tgis_message_interface()
    supported_stds_types = {
        "strds": "strds",
        "rast": "strds",
        "raster": "strds",
        "str3ds": "str3ds",
        "raster3d": "str3ds",
        "rast3d": "str3ds",
        "raster_3d": "str3ds",
        "stvds": "stvds",
        "vect": "stvds",
        "vector": "stvds",
    }
    if stds_type not in supported_stds_types:
        msgr.fatal(_("Unknown type: %s") % (type))
    return dataset_factory(supported_stds_types[stds_type], stds_id)]
```

File: python/grass/temporal/open_stds.py [hunk 10]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+        id = name
+
+    if type in {"strds", "rast", "raster"}:
+        sp = dataset_factory("strds", id)
+    elif type in {"str3ds", "raster3d", "rast3d", "raster_3d"}:
+        sp = dataset_factory("str3ds", id)
+    elif type in {"stvds", "vect", "vector"}:
+        sp = dataset_factory("stvds", id)
+    else:
+        msgr.fatal(_("Unknown type: %s") % (type))
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: See my comment below about code duplication. This is also part of the existing functions...

File: python/grass/temporal/open_stds.py [hunk 11]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+
+    return sp
+
+
+###############################################################################
+
+
+def check_open_output_stds(
+    name, type, dbif=None, overwrite: bool = False, extend: bool = False
+):
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: Also here, please add annotations.

File: python/grass/temporal/open_stds.py [hunk 12]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+    :param dbif: The temporal database interface to be used
+    :param overwrite: Flag to allow overwriting existing dataset
+    :param extend: Flag to extend an existing dataset if it exists
+
+    This function will raise a FatalError in case of an error.
+    """
+    mapset = get_current_mapset()
+    msgr = get_tgis_message_interface()
+
+    if name.find("@") < 0:
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: use `"@" in name`  instead of ` find`

File: python/grass/temporal/open_stds.py [hunk 13]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+        id = name + "@" + mapset
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: Use f-string

File: python/grass/temporal/open_stds.py [hunk 14]
```diff
@@ -135,14 +134,160 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+    :param overwrite: Flag to allow overwriting
+    :param extend: Flag to extend an existing dataset if it exists
+
+    :return: The opened or created space time dataset
+    """
+    dbif, connection_state_changed = init_dbif(dbif)
+    msgr = get_tgis_message_interface()
+
+    mapset = get_current_mapset()
+    id = name + "@" + mapset if name.find("@") < 0 else name
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: There are several variants of this line in the code. Maybe also suitable for a helper function `ensure_id(name)` .

**Review by ninsbl — CHANGES_REQUESTED** [review]
Please keep in mind that the code in ` open_stds` pre-dates Python 3 and has not been re-factored after since.
We should not do re-factoring as part of this PR, but don`t repeat outdated pattern in stuff you add....

File: python/grass/temporal/open_stds.py (hunk 9 above)
Comment (Sourish-spc) [reply python/grass/temporal/open_stds.py:357]: Added _get_stds helper function to avoid repeating the type mapping block, and _ensure_id helper to avoid repeating the id construction pattern. Kept them private (_ prefix) since they're only used internally in open_stds.py happy to make them public if needed.

File: python/grass/temporal/open_stds.py (hunk 11 above)
Comment (Sourish-spc) [reply python/grass/temporal/open_stds.py]: Added type annotations to all parameters including dbif: SQLDatabaseInterfaceConnection | None = None.

File: python/grass/temporal/open_stds.py [hunk 15]
```diff
@@ -135,14 +146,200 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
 
     dbif, connection_state_changed = init_dbif(dbif)
 
-    if sp.is_in_db(dbif) and overwrite is False:
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py:138]: The old syntax is preferable. See: https://docs.astral.sh/ruff/rules/collapsible-if/

File: python/grass/temporal/open_stds.py [hunk 16]
```diff
@@ -135,14 +146,200 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+                _("Space time datasets can only be created in the current mapset")
+            )
+
+    id = _ensure_id(name)
+    sp = _get_stds(id, type)
+
+    dbif, connection_state_changed = init_dbif(dbif)
+    output_exists = sp.is_in_db(dbif)
+
+    if output_exists:
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: See the comment above about collapsible-if...

File: python/grass/temporal/open_stds.py [hunk 17]
```diff
@@ -135,14 +146,200 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+    :param dbif: The temporal database interface to be used
+    :param overwrite: Flag to allow overwriting existing dataset
+    :param extend: Flag to extend an existing dataset if it exists
+    :return: True if the output dataset already exists, False otherwise
+
+    This function will raise a FatalError in case of an error.
+    """
+    msgr = get_tgis_message_interface()
+
+    if "@" in name:
```
Comment (ninsbl) [inline python/grass/temporal/open_stds.py]: You can skip this check if you `_ensure-id` before.
If ensure_id is a public function and used in `t.rast.aggregate`, the new functions you add here could expect an _id_ as input and those checks are not needed at all. Yet, they are insignificant and may be safer to double-check. No strong opinion from my side on this...

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py [hunk 18]
```diff
@@ -237,6 +237,72 @@ def test_aggregation_3months(self):
...
+        # (avoids hardcoding year or map count — works for any time range)
+        extended_map_list = extended_maps.strip().split(os.linesep)
+        b_ext_maps = [m for m in extended_map_list if m.startswith("b_ext")]
+        self.assertTrue(
+            len(b_ext_maps) > 0,
+            "No maps with basename b_ext found in extended STRDS",
+        )
+
+        # Check original maps are still present (STRDS was not overwritten)
+        for original_map in original_maps.strip().split(os.linesep):
```
Comment (ninsbl) [inline temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]: Maybe use all() or a set() based check and only one assertion...

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py [hunk 19]
```diff
@@ -237,6 +237,72 @@ def test_aggregation_3months(self):
...
+            self.assertIn(original_map, extended_maps)
+
+        # Check total map count increased (original 3 + extended 3 = 6)
+        info = SimpleModule("t.info", flags="g", input="B")
```
Comment (ninsbl) [inline temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]: You get the number of maps from t.rast.list. No need to run t.info in addition for this check. If you want to check that the STRDS history is updated properly (command added to comments) you need to.info...

**Review by ninsbl — CHANGES_REQUESTED** [review]
Now we are getting somewhere. Almost done. I just had some hopefully final, minor comments.
Once those are addressed, this can be approved as far as I am concerned. I would appreciate a second opinion, especially on having the new helper function public or private...

File: python/grass/temporal/open_stds.py (hunk 15 above)
Comment (Sourish-spc) [reply python/grass/temporal/open_stds.py:138]: Fixed: collapsed nested if statements into single conditions.

File: python/grass/temporal/open_stds.py (hunk 17 above)
Comment (Sourish-spc) [reply python/grass/temporal/open_stds.py]: Removed the redundant mapset check

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py (hunk 19 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]: Removed the t.info call, map count is now derived directly from t.rast.list output using len().

File: python/grass/temporal/open_stds.py (hunk 16 above)
Comment (Sourish-spc) [reply python/grass/temporal/open_stds.py]: Fixed: collapsed nested if statements into single conditions.

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py (hunk 18 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]: Replaced the loop with a single set.issubset() assertion.

**Comment by Sourish-spc** [comment]
Note: test_aggregation_relative.py was already failing before this PR's changes, verified by running the tests on the original code via git stash. This appears to be a pre-existing issue unrelated to the -e flag changes."

File: temporal/t.rast.aggregate/t.rast.aggregate.py [hunk 20]
```diff
@@ -221,11 +233,17 @@ def main():
             sp.get_relative_time_unit(),
             dbif,
         )
-
+        # Refresh STRDS object to reflect newly registered maps
+        output_strds.select(dbif)
```
Comment (ninsbl) [inline temporal/t.rast.aggregate/t.rast.aggregate.py:237]: What is the reason to add this here? Was it a necessary change? And what does it accomplish?

File: temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py [hunk 21]
```diff
@@ -237,6 +237,69 @@ def test_aggregation_3months(self):
...
+            overwrite=True,
+        )
+
+        # Verify new maps were added
+        lister = SimpleModule("t.rast.list", input="B", columns="name", flags="u")
+        self.runModule(lister)
+        extended_maps = lister.outputs.stdout
+
+        # Check that at least one map with the new basename prefix exists
+        # (avoids hardcoding year or map count — works for any time range)
```
Comment (ninsbl) [inline temporal/t.rast.aggregate/testsuite/test_aggregation_absolute.py]:
```suggestion
```
Year is hardcoded earlier anyway and the number of maps should be fixed too. It is probably OK not to check it here again, but the comment can be removed.

**Review by ninsbl — CHANGES_REQUESTED** [review]
Two minor and hopefully final comments from my side. Looks otherwise good. I would appreciate a second opinion though, esp. regarding the library parts...

File: temporal/t.rast.aggregate/t.rast.aggregate.py (hunk 20 above)
Comment (Sourish-spc) [reply temporal/t.rast.aggregate/t.rast.aggregate.py:237]: Yes, it was necessary to fix a bug. Without it, after register_map_object_list adds the new maps into the existing STRDS, the in-memory object still holds the old state (3 maps). select(dbif) refreshes it from the database so metadata.update records the correct map count (6). Removing it causes the extend test to fail.

**Review by ninsbl — APPROVED** [review]
Thanks for addressing my comments. @wenzeslaus What do you think? Ready to merge?

**Comment by Sourish-spc** [comment]
Any updates on this? Happy to address any further feedback.

**Review by ninsbl — DISMISSED** [review]
Looks good to me. @petrasovaa could you double-check the test and library code? Would be good to get this merged, so the functions can be reused...

File: python/grass/temporal/open_stds.py [hunk 22]
```diff
@@ -135,14 +146,191 @@ def check_new_stds(name, type, dbif=None, overwrite: bool = False):
...
+
+    :return: The opened or created space time dataset
+    """
+    dbif, connection_state_changed = init_dbif(dbif)
+    msgr = get_tgis_message_interface()
+
+    id = _ensure_id(name)
+    sp = _get_stds(id, type)
+
+    if extend and sp.is_in_db(dbif):
```
Comment (petrasovaa) [inline python/grass/temporal/open_stds.py:330]: Should this be using `check_new_stds` somewhere?

Comment (petrasovaa) [reply python/grass/temporal/open_stds.py:330]: Otherwise check you are writing/modifying data in current mapset is missing.

**Comment by petrasovaa** [comment]
I think the tests don't quite cover, e.g. -e and stds doesn't exist, extending with stds with incompatible temporal type...

File: python/grass/temporal/open_stds.py (hunk 22 above)
Comment (Sourish-spc) [reply python/grass/temporal/open_stds.py:330]: Addressed your review comments:

Added mapset ownership check when extending
Added temporal type compatibility check early in check_open_output_stds() before processing starts
Added fatal error when -e is used but output STRDS does not exist
Added two new test cases covering both edge cases
